neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
-from neuraxle.base import ResumableStepMixin, BaseStep, DataContainer, ListDataContainer, DEFAULT_CACHE_FOLDER, \
-    ExecutionContext
+from neuraxle.base import ResumableMixin, BaseStep, DataContainer, ListDataContainer, ExecutionContext, \
+    ExecutionMode
 
 
-class BaseCheckpointStep(ResumableStepMixin, BaseStep):
-    """
-    Base class for a checkpoint step that can persists the received data inputs, and expected_outputs
-    to eventually be able to load them using the checkpoint pipeline runner.
+class BaseCheckpointer(ResumableMixin):
     """
+    Base class to implement a step checkpoint or data container checkpoint.
 
-    def __init__(self, force_checkpoint_name: str = None):
-        ResumableStepMixin.__init__(self)
-        BaseStep.__init__(self)
-        self.force_checkpoint_name = force_checkpoint_name
-
-    def handle_fit(self, data_container: DataContainer, context: ExecutionContext) -> ('BaseStep', DataContainer):
-        data_container = self._handle_any(context, data_container)
-        return self, data_container
+    Checkpoint step uses many BaseCheckpointer to checkpoint both data container checkpoints, and step checkpoints.
 
-    def handle_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
-        data_container = self._handle_any(context, data_container)
-        return data_container
+    BaseCheckpointer has an execution mode so there could be different checkpoints for each execution mode (fit, fit_transform or transform).
+    """
+    def __init__(
+            self,
+            execution_mode: ExecutionMode
+    ):
+        self.execution_mode = execution_mode
 
-    def handle_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> ('BaseStep', DataContainer):
-        data_container = self._handle_any(context, data_container)
-        return self, data_container
+    def is_for_execution_mode(self, execution_mode) -> bool:
+        """
+        Returns true if the checkpointer should be used with the given execution mode.
 
-    def _handle_any(self, context, data_container):
-        self.set_checkpoint_path(context.get_path())
-        data_container: DataContainer = self.save_checkpoint(data_container)
+        :param execution_mode: execution mode (fit, fit_transform, or transform)
+        :return: if the checkpointer should be used
+        :rtype: bool
+        """
+        if execution_mode == ExecutionMode.FIT:
+            return self.execution_mode == ExecutionMode.FIT or \
+                   self.execution_mode == ExecutionMode.FIT_OR_FIT_TRANSFORM
 
-        self.save_checkpoint(data_container)
-        context.save_all_unsaved()
+        if execution_mode == ExecutionMode.FIT_TRANSFORM:
+            return self.execution_mode == ExecutionMode.FIT_TRANSFORM or \
+                   self.execution_mode == ExecutionMode.FIT_OR_FIT_TRANSFORM
 
-        return data_container
+        if execution_mode == ExecutionMode.FIT:
```

Comment:
Did you mean `execution_mode == ExecutionMode.TRANSFORM` here?

Reply:
yes